### ISSUE TITLE
Failed to check ruleset - Invalid index

### ISSUE THREAD

**Issue by grumpper** [terraform-linters/tflint-ruleset-aws#563]
I have the following resource declared as follows:

```
resource "aws_accessanalyzer_analyzer" "access_analyzer_instance" {
  count         = var.enable ? 1 : 0
  analyzer_name = "${var.resources_name_prefix["iam_access_analyzer"]}-lz"
  type          = var.type
  tags = {
      Name = "${var.resources_name_prefix["iam_access_analyzer"]}-lz"
  }
}
```

The `resources_name_prefix` variable is declared as `map(string)` with default value `{}`.
When checking with tflint I get the following error:

```
Failed to check ruleset; failed to check "aws_accessanalyzer_analyzer_invalid_analyzer_name" rule: terraform/main.tf:3,47-70: Invalid index; The given key does not identify an element in this collection value.
```

If I modify the code to be `analyzer_name = "${var.resources_name_prefix.iam_access_analyzer}-lz"` i get this error:

```
Failed to check ruleset; failed to check "aws_accessanalyzer_analyzer_invalid_analyzer_name" rule: terraform/main.tf:3,47-67: Missing map element; This map does not have an element with the key "iam_access_analyzer".
```

If I however modify the code to be `analyzer_name = "${lookup(var.resources_name_prefix, "iam_access_analyzer", "iaa")}-lz"` tflint makes a successful checks without throwing errors... Which is good except that the same expression used for specifying `analyzer_name` is also used to specify the `Name` tag in the `tags` block of the resource... So it looks like the we got problems with maps not having their keys immediately known only half of the time :)

And by the way tflint/aws should not be concerned with the presence or lack of keys in my map variables to begin with...

So I definetly think this is some kind of bug.
Even more so since `terraform validate` returns that the configuration is valid.

tflit version used 0.48
aws plugin used 0.27

**Comment by bendrucker** [comment]
This isn't specific to this rule/ruleset nor is it a bug. If you look up a map key that does not exist, that's an error, whether you use dot syntax or square brackets. 

A map is only suitable for key/value collections in which the keys are unknown, and typically the values are all of the same type. If the keys are known, an `object` is better.

> And by the way tflint/aws should not be concerned with the presence or lack of keys in my map variables to begin with...

https://github.com/terraform-linters/tflint/blob/master/docs/user-guide/config.md#variables

Your default value is not valid and the expression cannot be evaluated. The `validate` command is more lenient but in this case you are explicitly running a rule that needs to evaluate this expression to decide if `analyzer_name`'s value is valid. 

> So it looks like the we got problems with maps not having their keys immediately known

The keys _are_ known. They're empty, since that's the `default` you've set. You'll either need to provide a valid default or set the variable to valid value for the expressions TFLint needs to evaluate for your enabled rules.

**Comment by grumpper** [comment]
Not sure why this issue was so quickly closed...

Here is my response about its closure justification:

> This isn't specific to this rule/ruleset nor is it a bug. If you look up a map key that does not exist, that's an error, whether you use dot syntax or square brackets.

This is valid but when you go through the whole init/plan/apply ordeal which tflint does not do. 

> Your default value is not valid and the expression cannot be evaluated. The validate command is more lenient but in this case you are explicitly running a rule that needs to evaluate this expression to decide if analyzer_name's value is valid.

1. Having map variable with default value `{}` is perfectly valid.
2. tflint evaluating expressions need to know variables' values now? how come?

And to reiterate again. If what you are describing here is through and the problem is with how my variable is defined and used then the same problem should appear in the declaration of the `Name` tag (which is absolutely the same as the problematic declaration). However it looks like I am handling variables "incorrectly" only the first time (for the analyzer_name) but for the tag's value it's totally ok? :)

To give you another example with simpler code:

```
provider "aws" {
  region = "eu-west-1"
}

variable "sample_map" {
  type = map(string)
  default = {}
}

output "random_output" {
  value = "${var.sample_map["random_key"]}-test"
}

resource "aws_s3_bucket" "sample_bucket" {
  bucket = "${var.sample_map["random_key"]}-bucket"
}
```

tflint is completely fine with the above:

```
2 issue(s) found:

Warning: terraform "required_version" attribute is required (terraform_required_version)

  on  line 0:
   (source code not available)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.4.0/docs/rules/terraform_required_version.md

Warning: Missing version constraint for provider "aws" in `required_providers` (terraform_required_providers)

  on main.tf line 14:
  14: resource "aws_s3_bucket" "sample_bucket" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.4.0/docs/rules/terraform_required_providers.md
```

So where is the complaint here that a ruleset is failed to be checked due to invalid index? :)
Absolutely the same logic and approach yields different outcome in terms of whether an expression needs to be evaluated...

**Comment by bendrucker** [comment]
> Not sure why this issue was so quickly closed...

I'm confident in my explanation and that this is expected/intentional behavior. Happy to explain further, but the issue's status reflects that.

> Having map variable with default value `{}` is perfectly valid.

It is not a question of whether it is a valid value, meaning whether the static expression in `default` is valid. It surely is. The question is whether that default value, if used because no explicit value is set, is valid for the module. And in your case it isn't, you have expressions elsewhere in the module that assume that map will have a certain key. The default value is not usable in those expressions. For that default to be valid you need to provide fallbacks in the expression for when the expected key is unavailable, either using `lookup` or `try`. 

> Absolutely the same logic ... yields different outcome in terms of whether an expression needs to be evaluated...

Well yes, but the given example working with the default configuration is the intended behavior. The configuration is not valid, but because the outer structure is valid and the issue is in an expression, you don't get an error unless you enable a rule that has to evaluate the bad expression. Enable any rule that does and you get the same error:

```sh
tflint --only aws_s3_bucket_name
```

```
Failed to check ruleset; failed to check "aws_s3_bucket_name" rule: main.tf:15,29-43: Invalid index; The given key does not identify an element in this collection value.
```

This is intended behavior. Expressions that are _unknown_ are ignored. Expressions that are _known_ but invalid are returned as errors.

**Comment by bendrucker** [comment]
TFLint sits somewhere in between `terraform validate` and a full `apply`. They are complimentary tools. Ultimately what defines a valid configuration is what can pass `apply`. Support for variable values is a gap in `validate` that TFLint tries to fill.
